### PR TITLE
#104 Run tests for the user creating the PR

### LINKED ISSUE #103: Fix tests (richardstartin)
Hey @richardstartin, @AlexanderYastrebov added one more test for 10K distinct keys which fails with your impl. Could you take a look? Thx!

```
./test.sh richardstartin
Validating calculate_average_richardstartin.sh -- src/test/resources/samples/measurements-1.txt

real	0m0.077s
user	0m0.077s
sys	0m0.019s
Validating calculate_average_richardstartin.sh -- src/test/resources/samples/measurements-10.txt

real	0m0.065s
user	0m0.076s
sys	0m0.014s
Validating calculate_average_richardstartin.sh -- src/test/resources/samples/measurements-10000-unique-keys.txt
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException
	at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
	at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:542)
	at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:567)
	at java.base/java.util.concurrent.ForkJoinTask.join(ForkJoinTask.java:653)
	at dev.morling.onebrc.CalculateAverage_richardstartin.main(CalculateAverage_richardstartin.java:391)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 603 out of bounds for length 600
	at dev.morling.onebrc.CalculateAverage_richardstartin$Page.update(CalculateAverage_richardstartin.java:254)
	at dev.morling.onebrc.CalculateAverage_richardstartin$AggregationTask.computeSlice(CalculateAverage_richardstartin.java:307)
	at dev.morling.onebrc.CalculateAverage_richardstartin$AggregationTask.compute(CalculateAverage_richardstartin.java:339)
	at dev.morling.onebrc.CalculateAverage_richardstartin$AggregationTask.compute(CalculateAverage_richardstartin.java:276)
	at java.base/java.util.concurrent.RecursiveTask.exec(RecursiveTask.java:110)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
```

### LINKED ISSUE #104: Run test suite as part of PR build
Provided folks stick to the requested convention and name their launch script after their user name, we should be able to run the _tests.sh_ script as part of PR builds and fail them accordingly.

CC @AlexanderYastrebov, @filiphr

### PR DESCRIPTION
**Pull request by filiphr**
Adding a step in the GitHub action to test the submission of the user submitting the PR

Fixes #104

@gunnarmorling please wait for merging, I want to also push some failures to see how it looks like.

cc @AlexanderYastrebov

### CODE REVIEW

**Comment by filiphr** [comment]
A failure would look like

image

and then if you go into details you'll see

image

**Comment by filiphr** [comment]
I've also added a check to validate that the file exists. If a file does not exists it will have an error like

image

**Comment by filiphr** [comment]
@gunnarmorling this is now ready for merging. I've removed the commits that were causing failures.

File: test.sh [hunk 1]
```diff
@@ -22,12 +22,18 @@ if [ -z "$1" ]; then
...
 fi
 
+CALCULATE_FILE="./calculate_average_$1.sh"
+if [ ! -f "$CALCULATE_FILE" ]; then
+    echo "$CALCULATE_FILE does not exist."
+    exit 1
+fi
+
 for sample in $(ls src/test/resources/samples/*.txt); do
   echo "Validating calculate_average_$1.sh -- $sample"
```
Comment (AlexanderYastrebov) [inline test.sh:32]: I guess we can change this to `echo "Validating $1 -- $sample"` to make it a bit cleaner

File: .github/workflows/maven.yml [hunk 2]
```diff
@@ -47,3 +47,7 @@ jobs:
 
       - name: 'Build project'
         run: mvn -B clean verify -Pci
+
+      - name: 'Test submission'
+        run: ./test.sh ${{ github.event.pull_request.user.login }}
```
Comment (AlexanderYastrebov) [inline .github/workflows/maven.yml:52]: I am not sure this will work for all submissions, e.g. some use sdk or graal

**Comment by AlexanderYastrebov** [comment]
I thin PR description fixes wrong ticket, should be #104

**Comment by filiphr** [comment]
>I thin PR description fixes wrong ticket, should be https://github.com/gunnarmorling/1brc/issues/104

good point. I've updated that

File: test.sh (hunk 1 above)
Comment (filiphr) [reply test.sh:32]: Sounds OK to me. I'll change it.

File: .github/workflows/maven.yml (hunk 2 above)
Comment (filiphr) [reply .github/workflows/maven.yml:52]: Indeed it might not work. I'll see if we can use https://github.com/sdkman/sdkman-action/tree/main

Comment (gunnarmorling) [reply .github/workflows/maven.yml:52]: Yeah, I think this would be good. Wanna add it to this PR?

Comment (filiphr) [reply .github/workflows/maven.yml:58]: Was already working on it, but I am travelling today so a bit slow :)

**Comment by filiphr** [comment]
I've squashed all the commits and I've added SDKMAN as an action.

For this to work on the CI people will need to do something like:

```
source "$HOME/.sdkman/bin/sdkman-init.sh"
sdk install java 21.0.1-graal 1>&2
sdk use java 21.0.1-graal 1>&2
```

The `sdk install` is needed because the CI will not have the needed JDK. If the specified JDK is installed then `sdk install` is a noop (won't faill).


-----

Aside from this approach, I have also been looking into an alternative. However, it would require a change in all scripts.

The calculate script can look like

```sh
function java_to_use() {
    echo "21.0.1-graal"
}

function run_submission() {
  JAVA_OPTS=""
  time java $JAVA_OPTS --class-path target/average-1.0.0-SNAPSHOT.jar dev.morling.onebrc.CalculateAverage_<username>
}
```
The `java_to_use` is optional and if omitted then the default one is used. We would then have a `helper.sh` which looks like

```sh
source $HOME/.sdkman/bin/sdkman-init.sh

function sdk_use_java() {
  if [ -z "$1" ]; then
    echo "Usage: sdk_use_java <candidate>"
    exit 1
  fi
  java_to_use="$1"
  sdk install java ${java_to_use} 1>&2
  sdk use java ${java_to_use} 1>&2
}

if [ -n "$(LC_ALL=C type -t java_to_use)" ] && [ "$(LC_ALL=C type -t java_to_use)" = function ]; then
  sdk_java=$(java_to_use)
  sdk_use_java ${sdk_java}
fi

java -version
```

the `evaluate.sh` will need to then have before the loop

```sh
source ${CALCULATE_FILE}
source ./helper.sh
```

and in the loop the diff would be

```sh
diff <(run_submission) ${sample%.txt}.out
```

Similar chnage would be needed for the `evaluate.sh` and we would most likelt need to add a `run.sh` that would then run the `calculate` as well.

Let me know what you think about this idea.

File: test.sh (hunk 1 above)
Comment (filiphr) [reply test.sh:32]: I forgot to change this. I think we can merge without it, and we can change it later.

File: .github/workflows/maven.yml (hunk 2 above)
Comment (AlexanderYastrebov) [reply .github/workflows/maven.yml:57]: There is a related ticket https://github.com/gunnarmorling/1brc/issues/150

**Comment by gunnarmorling** [comment]
Could we somehow pre-install all the (currently) used distributions?

**Comment by gunnarmorling** [comment]
> Could we somehow pre-install all the (currently) used distributions?

Any thoughts on this, @filiphr? Or should we close it for now?

**Comment by filiphr** [comment]
Yes we can do that. I've been away with a really bad internet connection. I'm back on Sunday, will be able to look into it then

**Comment by gunnarmorling** [comment]
Ok, closing for now. Thx for looking into this, @filiphr!

**Comment by gunnarmorling** [comment]
Gave this another try, see https://github.com/gunnarmorling/1brc/pull/535.

**Pull request closed without merging**